// app/updates.js
var nodemailer = require('nodemailer');
var fs = require('fs');

var Updates = JSON.parse(fs.readFileSync('config.json', 'utf8')).updates;

// create reusable transporter object using SMTP transport
var transporter = nodemailer.createTransport({
    service: 'Zoho',
    auth: {
        user: "updates@dallen.io",
        pass: Updates.pass
    }
});

var update = {
  sendupdate: function(message){
      transporter.sendMail({
          from: 'Updates <updates@dallen.io>',
          to: Updates.receivers,
          subject: '',
          text: message
      }, function(error, info){
        if(error){
            return console.log(error);
        }
        console.log('Update sent: ' + info.response);
    })
  },
  sendMail: function(message, subj, to){
      transporter.sendMail({
          from: 'Updates <updates@dallen.io>',
          to: to,
          subject: subj,
          text: message
      }, function(error, info){
        if(error){
            return console.log(error);
        }
        console.log('Email sent: ' + info.response);
    })
  }
};

module.exports = update;
