// app/user.js
var mysql = require('mysql');
var debug = require('debug')('DallenJS:server');
var fs = require('fs');

var dbconnection = mysql.createConnection(GLOBAL.ConnStr);

var user = {
  getUsers: function(callback){
      var q = "select username, email, user_type from users";
      dbconnection.query(q, function(err, rows) {
          if(err){
            res.status(404);
            res.render('error', {
              message: "Not Found",
              error: {status: "404"}
            });
          }else{
            callback(rows);
          }
      });
  },
  addUser: function(profile, username){
    var q = "insert";
    dbconnection.query(q, function(err, rows) {
        if(err){
          res.status(404);
          res.render('error', {
            message: "Not Found",
            error: {status: "404"}
          });
        }else{
          callback(rows);
        }
    });
  },
  setUserType: function(username, user_type){
    var q = "update";
    dbconnection.query(q, function(err, rows) {
        if(err){
          res.status(404);
          res.render('error', {
            message: "Not Found",
            error: {status: "404"}
          });
        }else{
          callback(rows);
        }
    });
  },
  connect: function(){
    dbconnection.connect(function(err) {
      if (err) {
        console.error('error connecting: ' + err.stack);
        return;
      }
      console.log('connected as id ' + dbconnection.threadId);
    });
  }
};

module.exports = user;
